Preserve expanded broadcasts in reshape

jacobhinkle opened the pull request:
Currently, we always "realize" expanded broadcasts (replace them with `Iteration` domains) during reshape if they appear in a split or merge transform. This is not always necessary, for example we never need to do this for a `SplitTransform`, and we can avoid it for a `MergeTransform` if both inner and outer IDs are `Broadcast`. This PR makes that a reality.

Fixes #1126.

jacobhinkle commented:
!build

jacobhinkle commented:
!build --diff

jacobhinkle commented:
!build

jacobhinkle commented:
!build --diff

wujingyue commented on csrc/transform_view.cpp:
```diff
@@ -137,18 +138,23 @@ class ViewTransform : public Transform {
+    if (realize_expand && id->hasExpandedExtent()) {
+      // Convert an expanded broadcast to Iteration (resolve the broadcast)
+      iter_type = IterType::Iteration;
+      extent = id->getMaybeExpandedExtent();
```
```suggestion
      extent = expanded_extent;
```

jacobhinkle replied:
I prefer to leave this explicit for clarity since I'm changing the value of `expanded_extent` in the next line.

jacobhinkle replied:
Actually on second thought I like this suggestion better.

wujingyue commented on csrc/transform_view.cpp:
```diff
@@ -197,24 +203,43 @@ class MergeTransform final : public ViewTransform {
+    // are broadcasts. Otherwise, we resolve the broadcast and produce an
+    // Iteration output domain.
+    bool realize_expand = !(outer_id->isBroadcast() && inner_id->isBroadcast());
+    auto iter_type = realize_expand ? IterType::Iteration : IterType::Broadcast;
```
Move this closer to where it's first used (line 240) or simply inline it (line 240 appears to be the only use of `iter_type`).

wujingyue commented on csrc/transform_view.cpp:
```diff
@@ -269,27 +294,43 @@ class SplitTransform final : public ViewTransform {
-    Val* remainder = ceilDiv(id->extent(), factor);
+    Val* remainder = ceilDiv(id->getMaybeExpandedExtent(), factor);
+
+    auto outer_extent = factor;
+    Val* outer_expanded_extent = nullptr;
+    auto inner_extent = remainder;
+    Val* inner_expanded_extent = nullptr;
+    if (id->hasExpandedExtent()) {
+      outer_extent = FusionGuard::getCurFusion()->oneVal();
+      outer_expanded_extent = factor;
+      inner_extent = FusionGuard::getCurFusion()->oneVal();
+      inner_expanded_extent = remainder;
+    }
```
Instead of assigning an initial value and later maybe replacing it, I think it's clearer to just assign the right value for once. 

```suggestion
    Val* outer_extent, *outer_expanded_extent;
    Val* inner_extent, *inner_expanded_extent;
    if (id->hasExpandedExtent()) {
      outer_extent = FusionGuard::getCurFusion()->oneVal();
      outer_expanded_extent = factor;
      inner_extent = FusionGuard::getCurFusion()->oneVal();
      inner_expanded_extent = remainder;
    } else {
      outer_extent = factor;
      outer_expanded_extent = nullptr;
      inner_extent = remainder;
      inner_expanded_extent = nullptr;
    }
```

wujingyue commented on test/test_gpu_view.cpp:
```diff
@@ -2478,4 +2478,53 @@ TEST_F(GpuViewTest, ExpandedBroadcast) {
+}
+
+// Reverse of SplitExpandedBroadcast
+TEST_F(GpuViewTest, MergeExpandedBroadcast) {
```
Also test merging a broadcast and a non-broadcast?

jacobhinkle replied:
Done.

wujingyue reviewed (approved):
Nice work! LGTM other than some nit picks.

jacobhinkle commented on csrc/transform_view.cpp:
```diff
@@ -137,18 +138,23 @@ class ViewTransform : public Transform {
     auto root_domain_pos = std::distance(root_domain.begin(), root_domain_it);
 
-    bool is_expanded_dim = id->hasExpandedExtent();
+    auto iter_type = id->getIterType();
```
In general I agree, but in this case it's an `IterType` and the method is named `getIterType` so it's not the most mysterious case. :laughing:

tfogal replied:
I want to second Jingyue's comment here.

"Not clear from context" to me in practice means "not already in the line in question". I can see why this case feels like a gray area, but to the uninitiated it doesn't feel that way: there's no understanding that `IterType` is an actual type. Furthermore, without any kind of enforcement it's hard to trust a method name: even if it were `auto foo = bar->getString()` it's still ambiguous whether that returns a std::string or string_view or some custom string etc.

jacobhinkle replied:
Fair enough. I'll change it.

jacobhinkle commented:
> one real failure in `GpuViewTest.FusionExpandView2`:

I am not sure if this is related to the failure, but that test introduces something strange. During scheduling (not during the reshape transform done at definition), a merge is performed between an expanded broadcast and an `Iteration` domain. The result is an `Iteration` domain with an expanded extent. I hacked `IterDomain::toString()` to show what I mean here:
```
T3_l[ iblockIdx.x72{( ceilDiv(( ceilDiv(( ceilDiv(( 1 * ( 1 * 8 ) ), 128) ), 1) ), 1) ), EXPANDED_EXTENT=( ceilDiv(( ceilDiv(( ceilDiv(( 
3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) ), 1) )}, iUS73{1}, iS71{1}, ithreadIdx.x69{128} ] ca_pos( 4 )       
 root domain : (bS9{( (nvfuser_index_t)(12) ), extent=1}rf, iS8{8})                                                                      
  Outer split: bS9{( (nvfuser_index_t)(12) ), extent=1}rf by factor 3 -> bS10{3, extent=1}rf, bS11{( ceilDiv(( (nvfuser_index_t)(12) ), 3
) ), extent=1}rf, start offset: 0, stop offset: 0                                                                                        
 rfactor domain : (bS10{3, extent=1}rf, bS11{( ceilDiv(( (nvfuser_index_t)(12) ), 3) ), extent=1}rf, iS8{8})                             
 contiguity: n n t          
  Merge: bS11{( ceilDiv(( (nvfuser_index_t)(12) ), 3) ), extent=1}rf and iS8{8} -> iS66{( 1 * 8 ), EXPANDED_EXTENT=( ( ceilDiv(( (nvfuser
_index_t)(12) ), 3) ) * 8 )}                                        
  Merge: bS10{3, extent=1}rf and iS66{( 1 * 8 ), EXPANDED_EXTENT=( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 )} -> iS67{( 1 * ( 1 * 8
 ) ), EXPANDED_EXTENT=( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) )}                                                         
  Split: iS67{( 1 * ( 1 * 8 ) ), EXPANDED_EXTENT=( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) )} by factor 128 -> iS68{( ceilD
iv(( 1 * ( 1 * 8 ) ), 128) ), EXPANDED_EXTENT=( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) )}, ithreadIdx.x6
9{128}, start offset: 0, stop offset: 0                             
  Split: iS68{( ceilDiv(( 1 * ( 1 * 8 ) ), 128) ), EXPANDED_EXTENT=( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 1
28) )} by factor 1 -> iS70{( ceilDiv(( ceilDiv(( 1 * ( 1 * 8 ) ), 128) ), 1) ), EXPANDED_EXTENT=( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (
nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) )}, iS71{1}, start offset: 0, stop offset: 0                                            
  Split: iS70{( ceilDiv(( ceilDiv(( 1 * ( 1 * 8 ) ), 128) ), 1) ), EXPANDED_EXTENT=( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index
_t)(12) ), 3) ) * 8 ) ), 128) ), 1) )} by factor 1 -> iblockIdx.x72{( ceilDiv(( ceilDiv(( ceilDiv(( 1 * ( 1 * 8 ) ), 128) ), 1) ), 1) ), 
EXPANDED_EXTENT=( ceilDiv(( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) ), 1) )}, iUS73{1}, s
tart offset: 0, stop offset: 0                                                                                                           
 leaf domain : (iblockIdx.x72{( ceilDiv(( ceilDiv(( ceilDiv(( 1 * ( 1 * 8 ) ), 128) ), 1) ), 1) ), EXPANDED_EXTENT=( ceilDiv(( ceilDiv(( 
ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) ), 1) )}, iUS73{1}, iS71{1}, ithreadIdx.x69{128})
```
`T3_l` is the reshape result and this is printed from gdb where we see the error.

jacobhinkle commented:
I have a fix for this test failure but I'm stacking it on #1329 first.

jacobhinkle commented:
!build

naoyam reviewed (approved):
LGTM. Thanks for this improvement!

jacobhinkle commented:
!build

jacobhinkle commented:
> Maybe `SyncMap` and some other lowering machinery is just not used to having splits of `Broadcast` IDs yet. Looking into it...

SyncMap is failing because `useSameIndex` fails, because the parallelized indices between T2 and T3 (both BIDx and TIDx) are not being LOOP mapped. The transforms involved:
```
T2_l[ iblockIdx.x80{( ceilDiv(( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) ), 1) )}, iUS81{1}, iS79{1}, ithreadIdx.x77{128} ]
 root domain : (bS5{( (nvfuser_index_t)(12) )}, iS6{8})
 contiguity: n t
  Outer split: bS5{( (nvfuser_index_t)(12) )} by factor 3 -> bS28{3}, bS29{( ceilDiv(( (nvfuser_index_t)(12) ), 3) )}, start offset: 0, stop offset: 0
  Merge: bS29{( ceilDiv(( (nvfuser_index_t)(12) ), 3) )} and iS6{8} -> iS74{( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 )}
  Merge: bS28{3} and iS74{( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 )} -> iS75{( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) )}
  Split: iS75{( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) )} by factor 128 -> iS76{( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) )}, ithreadIdx.x77{128}, start offset: 0, stop offset: 0
  Split: iS76{( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) )} by factor 1 -> iS78{( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) )}, iS79{1}, start offset: 0, stop offset: 0
  Split: iS78{( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) )} by factor 1 -> iblockIdx.x80{( ceilDiv(( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) ), 1) )}, iUS81{1}, start offset: 0, stop offset: 0
 leaf domain : (iblockIdx.x80{( ceilDiv(( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) ), 1) )}, iUS81{1}, iS79{1}, ithread
Idx.x77{128})
T3_l[ iblockIdx.x72{( ceilDiv(( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) ), 1) )}, iUS73{1}, iS71{1}, ithreadIdx.x69{128} ] ca_pos( 4 )
 root domain : (bS9{( (nvfuser_index_t)(12) )}rf, iS8{8})
  Outer split: bS9{( (nvfuser_index_t)(12) )}rf by factor 3 -> bS10{3}rf, bS11{( ceilDiv(( (nvfuser_index_t)(12) ), 3) )}rf, start offset: 0, stop offset: 0
 rfactor domain : (bS10{3}rf, bS11{( ceilDiv(( (nvfuser_index_t)(12) ), 3) )}rf, iS8{8})
 contiguity: n n t
  Merge: bS11{( ceilDiv(( (nvfuser_index_t)(12) ), 3) )}rf and iS8{8} -> iS66{( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 )}
  Merge: bS10{3}rf and iS66{( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 )} -> iS67{( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) )}
  Split: iS67{( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) )} by factor 128 -> iS68{( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) )}, ithreadIdx.x69{128}, start offset: 0, stop offset: 0
  Split: iS68{( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) )} by factor 1 -> iS70{( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) )}, iS71{1}, start offset: 0, stop offset: 0
  Split: iS70{( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) )} by factor 1 -> iblockIdx.x72{( ceilDiv(( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) ), 1) )}, iUS73{1}, start offset: 0, stop offset: 0
 leaf domain : (iblockIdx.x72{( ceilDiv(( ceilDiv(( ceilDiv(( 3 * ( ( ceilDiv(( (nvfuser_index_t)(12) ), 3) ) * 8 ) ), 128) ), 1) ), 1) )}, iUS73{1}, iS71{1}, ithreadIdx.x69{128})
```

jacobhinkle commented:
In this case, the reshape is splitting an expanded dimension, so we could handle this the way we handle size-1 extents in the output of reshape: by squeezing the input dimension then appending a `BroadcastOp` after the reshape. In this case we'd also need an `ExpandOp`. It's slightly nuanced in this case, since sometimes the reshape will need to merge a broadcast/expanded IterDomain with an Iteration domain, and in those cases we cannot split it off into a separate op. I will look into adding this: we might need to track `expand_transforms_` as in addition to squeeze, broadcast, and view (merge/split) transforms in `AnalyzeViewTransformation`.

jacobhinkle commented on tests/cpp/test_gpu_view.cpp:
```diff
@@ -2312,4 +2312,103 @@ TEST_F(GpuViewTest, ExpandedBroadcast) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  TensorView* in = makeContigConcreteTensor({4, 5});
+  fusion.addInput(in);
+  TensorView* out = broadcast(in, {false, false, true});
+  out = expand(
+      out,
+      {IrBuilder::create<Val>(4),
+       IrBuilder::create<Val>(5),
+       IrBuilder::create<Val>(6)});
+  out = reshape(out, {4, 5, 6}, {4, 5, 2, 3});
+  fusion.addOutput(out);
```
This is what `schedulePointwise` gives us:
```
T5_l[ iblockIdx.x40{( ceilDiv(( ceilDiv(( ceilDiv(( 4 * ( 5 * ( 2 * ( ceilDiv(6, 2) ) ) ) ), 128) ), 1) ), 1) )}, iUS41{1}, iS39{1}, ithreadIdx.x37{128} ] ca_pos( 4 ) produce_pos( 4 )
 root domain : (iS8{4}, iS9{5}, bS11{1 ex 6}rf)
  Outer split: bS11{1 ex 6}rf by factor 2 -> bS12{1 ex 2}rf, bS13{1 ex ( ceilDiv(6, 2) )}rf, start offset: 0, stop offset: 0
 rfactor domain : (iS8{4}, iS9{5}, bS12{1 ex 2}rf, bS13{1 ex ( ceilDiv(6, 2) )}rf)
 contiguity: t t n n
  Merge: bS12{1 ex 2}rf and bS13{1 ex ( ceilDiv(6, 2) )}rf -> bS33{( 1 * 1 ) ex ( 2 * ( ceilDiv(6, 2) ) )}
  Merge: iS9{5} and bS33{( 1 * 1 ) ex ( 2 * ( ceilDiv(6, 2) ) )} -> iS34{( 5 * ( 2 * ( ceilDiv(6, 2) ) ) )}
  Merge: iS8{4} and iS34{( 5 * ( 2 * ( ceilDiv(6, 2) ) ) )} -> iS35{( 4 * ( 5 * ( 2 * ( ceilDiv(6, 2) ) ) ) )}
  Split: iS35{( 4 * ( 5 * ( 2 * ( ceilDiv(6, 2) ) ) ) )} by factor 128 -> iS36{( ceilDiv(( 4 * ( 5 * ( 2 * ( ceilDiv(6, 2) ) ) ) ), 128) )}, ithreadIdx.x37{128}, start offset: 0, stop offset: 0
  Split: iS36{( ceilDiv(( 4 * ( 5 * ( 2 * ( ceilDiv(6, 2) ) ) ) ), 128) )} by factor 1 -> iS38{( ceilDiv(( ceilDiv(( 4 * ( 5 * ( 2 * ( ceilDiv(6, 2) ) ) ) ), 128) ), 1) )}, iS39{1}, start offset: 0, stop offset: 0
  Split: iS38{( ceilDiv(( ceilDiv(( 4 * ( 5 * ( 2 * ( ceilDiv(6, 2) ) ) ) ), 128) ), 1) )} by factor 1 -> iblockIdx.x40{( ceilDiv(( ceilDiv(( ceilDiv(( 4 * ( 5 * ( 2 * ( ceilDiv(6, 2) ) ) ) ), 128) ), 1) ), 1) )}, iUS41{1}, start offset: 0, stop offset: 0
 leaf domain : (iblockIdx.x40{( ceilDiv(( ceilDiv(( ceilDiv(( 4 * ( 5 * ( 2 * ( ceilDiv(6, 2) ) ) ) ), 128) ), 1) ), 1) )}, iUS41{1}, iS39{1}, ithreadIdx.x37{128})
```
That looks OK to me but this is the CUDA kernel that's generated:
```c++
__global__ void nvfuser_none_f0_c0_r0_g0(Tensor<float, 2, 2> T0, Tensor<float, 4, 4> T3) {
  nvfuser_index_t i0;
  i0 = ((nvfuser_index_t)threadIdx.x) + (128LL * ((nvfuser_index_t)blockIdx.x));
  nvfuser_index_t i1;
  i1 = i0 % 30LL;
  if (((i0 < 120LL) && (i1 < 5LL))) {
    float T4[1LL];
    T4[0LL] = 0LL;
    T4[0LL]
       = T0[((5LL * (i0 / 30LL)) + i1)];
    float T1[1LL];
    T1[0LL]
       = T4[0LL];
    float T2[1LL];
    T2[0LL]
       = T1[0LL];
    float T5[1LL];
    T5[0LL]
       = T2[0LL];
    T3[i0]
       = T5[0LL];
  }
}
```
`i0` is the index into a 120-element vector, as if the [4, 5, 6] sized output was unexpanded. It is used to index the output store. The load and store are predicated such that `i0 % 30 < 5`, but that's not enough to guarantee a proper store to an array of size [4, 5]; for example, `i0 = 90` corresponds to `i1 = i0 % 30 == 0` so we will load element `T0[5 * (i0 / 30) + i1] = T0[15]` and write it to `T3[90]`, which is actually out of bounds assuming we have allocated the output properly as size 20. We don't see this prior to this PR since the output is unexpanded and the `i1 < 5` predicate is absent.

I think this might show that expanded dst indexing is not working properly, as I would expect us to use the same index for `T0` and `T3` here.

naoyam commented:
Could you please summarize what the current issues are?
